File: apps/desktop/src/desktop/project-directory-bridge.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * The typed desktop bridge for the project directory the user picks.
 *
 * This module belongs to the **desktop system layer**, not to the Runtime client and
 * not to the project read model. `src/runtime/` describes what a project *is*; this
 * file only learns *where the user pointed*, and it deliberately learns nothing else.
 *
 * ```text
 * renderer
 *   ↓ invoke(SELECT_PROJECT_DIRECTORY_COMMAND)   ← the only argument: none
 * Rust desktop system layer
 *   ↓ native directory picker
 * "C:\canx\projects\Vehicle_Bus"  |  null
 *   ↓ handed on verbatim, or reported as a typed failure
 * string | null
 * ```
 *
 * Four things this module refuses to do, each of them load-bearing:
 *
 * * **It does not name a directory.** There is no parameter to pass, so the renderer
 *   cannot pick a path the user never picked.
 * * **It does not decide what a project is.** Reading the project's definition file,
 *   validating its schema or checking that the directory holds a project at all
 *   belongs to the project read model — one authority, not two.
 * * **It does not touch the filesystem.** No `fs`, no `readFile`, no `existsSync`: a
 *   renderer that reads files directly would bypass the Runtime's project ownership.
 * * **It keeps no project state.** No `currentProject`, no module-level variable: the
 *   caller that knows which project it means passes the string along itself.
 */

/**
 * The Tauri command that opens the native directory picker.
 *
 * Spelled once here so the two halves of the bridge cannot drift apart.
 */
export const SELECT_PROJECT_DIRECTORY_COMMAND = "select_project_directory";

/**
 * The code reported when the bridge itself could not be reached.
 *
 * Distinct from every code the shell reports on purpose: those describe what the
 * *user's choice* was, and this one describes the absence of a bridge — which no
 * amount of choosing a different directory can fix.
 */
export const PROJECT_DIRECTORY_BRIDGE_UNAVAILABLE_CODE =
  "desktop.project_directory_bridge_unavailable";

/**
 * A typed failure that crossed the Tauri IPC boundary.
 *
 * Mirrors the `code` / `message` / `recoverable` contract the rest of CAN-X uses, so
 * a caller branches on the same three facts here as it does for an HTTP failure.
 * Messages never contain the directory the user picked.
 */
export class ProjectDirectoryBridgeError extends Error {
  /** Stable diagnostic code, for example `desktop.project_directory_not_a_directory`. */
  readonly code: string;
  /** Whether retrying with a different directory could reasonably succeed. */
  readonly recoverable: boolean;

  constructor(code: string, message: string, recoverable: boolean) {
    super(`${code}: ${message}`);
    this.name = "ProjectDirectoryBridgeError";
    this.code = code;
    this.recoverable = recoverable;
  }
}

/**
 * Open the native directory picker and resolve the directory the user chooses.
 *
 * Returns:
 * * the chosen directory as the shell reported it, character for character — this
 *   module normalises, resolves, expands or truncates nothing, because the string it
 *   returns is the string the caller will hand to the project read model;
 * * `null` when the user dismissed the picker. Cancel is control flow, never an
 *   error, and a caller checks one fact instead of guessing which rejection means
 *   "the user changed their mind".
 *
 * Throws:
 * * {@link ProjectDirectoryBridgeError} when the shell refused the choice — not a
 *   directory, unreadable, or a picker that failed to open. The `code` says which.
 * * {@link ProjectDirectoryBridgeError} with
 *   {@link PROJECT_DIRECTORY_BRIDGE_UNAVAILABLE_CODE} when the desktop shell is not
 *   there to answer at all — no shell, a serialization failure on the Rust side, or a
 *   rejection from a layer that has nothing to do with directory selection.
 * * a plain `Error` when the shell answered with something this contract does not
 *   describe. A payload that is neither a string nor `null` is a defect in the
 *   bridge, and it is reported as a contract error rather than being reshaped into a
 *   plausible path or mistaken for an unavailable shell.
 */
export async function selectProjectDirectory(): Promise<string | null> {
  let raw: unknown;
  try {
    raw = await invoke<string | null>(SELECT_PROJECT_DIRECTORY_COMMAND);
  } catch (cause) {
    throw toBridgeError(cause);
  }
  return readSelectedProjectDirectory(raw);
}

/**
 * Validate the shell's answer and project it onto `string | null`.
 *
 * The two legal shapes are checked positively: a string is returned as it arrived,
 * `null` means the user cancelled, and anything else is rejected. A payload that
 * carried an object — however plausibly named its fields — is not searched for a
 * directory to promote out of it.
 */
function readSelectedProjectDirectory(value: unknown): string | null {
  if (value === null) return null;
  if (typeof value !== "string") {
    throw new Error(
      "the project directory selection payload must be a string or null",
    );
  }
  return value;
}

/**
 * Turn whatever `invoke` rejected with into a {@link ProjectDirectoryBridgeError}.
 *
 * A rejection that already carries the three structured fields came from the Rust
 * bridge and is passed through intact. Anything else — an unavailable shell, a
 * serialization failure, a rejection from a layer that has nothing to do with this
 * selection — becomes the single "bridge unavailable" diagnosis rather than being
 * decorated with a code it never claimed.
 */
function toBridgeError(cause: unknown): ProjectDirectoryBridgeError {
  if (cause instanceof ProjectDirectoryBridgeError) return cause;
  const shaped = readBridgeErrorShape(cause);
  if (shaped !== null) {
    return new ProjectDirectoryBridgeError(shaped.code, shaped.message, shaped.recoverable);
  }
  return new ProjectDirectoryBridgeError(
    PROJECT_DIRECTORY_BRIDGE_UNAVAILABLE_CODE,
    "The desktop project directory bridge is unavailable.",
    false,
  );
}

function readBridgeErrorShape(
  value: unknown,
): { code: string; message: string; recoverable: boolean } | null {
  if (typeof value !== "object" || value === null) return null;
  const record = value as Record<string, unknown>;
  const code = record["code"];
  const message = record["message"];
  const recoverable = record["recoverable"];
  if (typeof code !== "string" || typeof message !== "string" || typeof recoverable !== "boolean") {
    return null;
  }
  return { code, message, recoverable };
}

File: apps/desktop/src-tauri/src/project_directory_bridge.rs
//! The desktop system layer's project-directory selection bridge.
//!
//! This module owns exactly one capability: turning a **user's explicit choice in a
//! native directory dialog** into the directory path that the Python Runtime already
//! treats as the input of its project-inspection contract. It is the filesystem half
//! of "open a project", and it exists so a renderer can name a project by *pointing at
//! it* without ever being able to read a path of its own choosing.
//!
//! ```text
//! User
//!   ↓
//! Native OS directory dialog        (trusted: opened by Rust, not by the renderer)
//!   ↓
//! Tauri trusted system layer
//!   ↓
//! the chosen directory path, as text
//!   ↓
//! typed TypeScript desktop bridge   (src/desktop/project-directory-bridge.ts)
//! ```
//!
//! Three properties are load-bearing, and each is enforced here rather than
//! documented and hoped for:
//!
//! * **No caller filesystem path, ever.** The only registered command,
//!   [`select_project_directory`], declares no argument a caller can supply. It opens
//!   the dialog itself and receives the chosen directory *internally*, from the OS.
//!   There is no `path`, no `directory` and no `project_path` for a renderer to
//!   populate, and no way to reach this function that would let one.
//! * **Selection is the whole authority.** This bridge decides nothing about what a
//!   project *is*: the directory the user points at is relayed as text and never
//!   inspected. Whether that directory is a CAN-X project — its layout, its manifest,
//!   its database — is the Runtime's decision, taken against the same path this bridge
//!   merely carries. The bridge proves no property of the selection; it transports it.
//! * **Cancel is control flow, not an error.** A dismissed dialog resolves to
//!   `Ok(None)`; the renderer sees `null`. Nothing downstream has to translate a
//!   "user changed their mind" signal out of an error channel.
//!
//! A failure never carries the selection. [`ProjectDirectoryBridgeError`] holds a code,
//! a human message and a recoverability flag; every message is static text, no
//! constructor accepts a path, and there is no field into which one could leak.

use std::path::Path;

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Runtime};
use tauri_plugin_dialog::{DialogExt, FilePath};

/// The name of the Tauri command that opens the native directory dialog.
///
/// Exported so that the architecture regression test can assert the registered
/// command surface without restating the name, and so the TypeScript bridge and the
/// Rust layer cannot drift apart on a string.
pub const SELECT_PROJECT_DIRECTORY_COMMAND: &str = "select_project_directory";

/// A typed failure from the desktop project-directory bridge.
///
/// Tauri IPC is an architecture boundary exactly like HTTP is, so a failure that
/// crosses it carries the same three structured fields the rest of CAN-X uses: a
/// stable `code` a caller can branch on, a human `message`, and whether retrying could
/// reasonably succeed.
///
/// Every message here is static text. A refusal must never become a rendering of the
/// directory that was refused, so no constructor accepts a path, and there is no field
/// into which one could leak.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProjectDirectoryBridgeError {
    /// Stable diagnostic code, for example `desktop.project_directory_dialog_failed`.
    pub code: String,
    /// Human-readable explanation. Never contains a filesystem path.
    pub message: String,
    /// Whether the caller can fix this by trying again.
    pub recoverable: bool,
}

impl ProjectDirectoryBridgeError {
    /// Build one failure from a code and a human message.
    fn of(code: &str, message: &str, recoverable: bool) -> Self {
        Self {
            code: code.to_owned(),
            message: message.to_owned(),
            recoverable,
        }
    }

    /// The native dialog itself could not be opened or did not answer.
    #[must_use]
    pub fn dialog_failed() -> Self {
        Self::of(
            "desktop.project_directory_dialog_failed",
            "The project directory dialog could not be opened.",
            true,
        )
    }

    /// The dialog answered with a location that is not a filesystem path.
    ///
    /// A directory picker on desktop hands back a path, but the dialog location type
    /// is shared with mobile content URIs; anything this bridge cannot turn into a
    /// path is refused here rather than guessed at.
    #[must_use]
    pub fn unavailable() -> Self {
        Self::of(
            "desktop.project_directory_unavailable",
            "The selected location could not be resolved to a filesystem path.",
            true,
        )
    }

    /// The chosen directory exists but its path cannot be represented as text.
    ///
    /// A path is not guaranteed to be valid Unicode. Rather than lossily rewriting the
    /// user's selection into something the Runtime could not open, the bridge refuses.
    #[must_use]
    pub fn not_representable() -> Self {
        Self::of(
            "desktop.project_directory_not_representable",
            "The selected directory path cannot be represented as text.",
            false,
        )
    }
}

/// Open the native directory dialog and return the directory the user chooses.
///
/// The one command this module registers. Its signature is the security invariant, not
/// merely a detail of it: **it declares no argument a caller can supply.** The
/// `AppHandle` is injected by Tauri from the running application, and there is nothing
/// else — so there is no way for a renderer to name a directory, and no way to reach
/// this function that would let one.
///
/// ```text
/// arguments: none
/// ```
///
/// Returns:
/// * `Ok(Some(path))` — the user chose a directory; `path` is exactly what the dialog
///   returned, as text. This bridge does **not** judge whether that directory is a
///   project: that decision belongs to the Runtime, against this same path.
/// * `Ok(None)` — the user dismissed the dialog. Cancel is normal control flow; it is
///   not reported as a failure, and the renderer receives `null`.
/// * `Err(error)` — the choice could not be turned into a path: a dialog that failed to
///   open, a location that is not a path, or a path that is not valid text.
///
/// The blocking dialog API is run on the async runtime's blocking pool, never on the
/// main thread: the dialog pumps its own message loop, and Tauri's event loop has to
/// stay free to serve it.
///
/// `pub(crate)` rather than `pub` on purpose. Tauri's command macro only exports its
/// generated wrapper macros to the crate root when the function is fully `pub`, and
/// that would move them *out of* this module — leaving `generate_handler!` in `lib.rs`
/// with a module path and no macro to resolve it against. The command is assembled by
/// this crate and reachable only through it; there is no reason for it to be part of
/// the library's public surface.
#[tauri::command]
pub(crate) async fn select_project_directory<R: Runtime>(
    app: AppHandle<R>,
) -> Result<Option<String>, ProjectDirectoryBridgeError> {
    let handle = app.clone();
    let picked =
        tauri::async_runtime::spawn_blocking(move || handle.dialog().file().blocking_pick_folder())
            .await
            .map_err(|_| ProjectDirectoryBridgeError::dialog_failed())?;

    resolve_selected_directory(picked)
}

/// Turn a dialog outcome into the value the bridge publishes.
///
/// **Not a command.** It is a plain function that happens to accept the dialog's own
/// [`FilePath`] because that is what resolving a selection requires; the value reaches
/// it from the dialog in the process above and from nowhere else. Keeping it a plain
/// function is what makes both halves of the contract — a dismissed dialog and a chosen
/// directory — testable without automating an OS dialog, and it is also what keeps
/// "receives a location" from ever becoming a renderer-facing capability.
///
/// ```text
/// None                       → Ok(None)                       (the user dismissed it)
/// Some(path)                 → Ok(Some(path as text))
/// Some(non-path location)    → Err(desktop.project_directory_unavailable)
/// Some(path that is not text)→ Err(desktop.project_directory_not_representable)
/// ```
fn resolve_selected_directory(
    picked: Option<FilePath>,
) -> Result<Option<String>, ProjectDirectoryBridgeError> {
    let Some(chosen) = picked else {
        return Ok(None);
    };

    // `FilePath` is a path on desktop and a content URI on mobile; this bridge is a
    // desktop bridge, and anything else is not something it can hand to the Runtime.
    let path = chosen
        .into_path()
        .map_err(|_| ProjectDirectoryBridgeError::unavailable())?;

    directory_text(&path).map(Some)
}

/// Render a chosen directory path as the text the caller receives.
///
/// A refusal rather than a lossy rewrite: a path that is not valid Unicode is not
/// silently replaced with one that is, because the caller would then hold a string
/// that names a directory the user never chose.
fn directory_text(path: &Path) -> Result<String, ProjectDirectoryBridgeError> {
    path.to_str()
        .map(str::to_owned)
        .ok_or_else(ProjectDirectoryBridgeError::not_representable)
}

#[cfg(test)]
mod tests;
